Postgres backend

espadolini opened the pull request:
Subordinate to the RFD at #28355; opening for review as I'm expecting only minor UX/config changes at this point.

This PR adds a backend implementation that uses PostgreSQL as storage. In addition to using fixed credentials, the implementation can also fetch Azure AD tokens, to log in to Azure Database for PostgreSQL database clusters.

espadolini committed:
Postgres backend

espadolini committed:
Use batch to mark read transactions as read only

espadolini committed:
Allow the backend_version creation to fail

espadolini committed:
clarify setupAndMigrate

espadolini committed:
simplify AzureBeforeConnect

espadolini committed:
Allow idempotent retries for reads

ravicious commented:
Should I ask someone with more knowledge in this area to review this? I'm afraid I'm able to give only a surface-level review, I have a pretty limited knowledge about the intricacies of our backend.

espadolini committed:
rm go.sum && go mod tidy

jakule commented on lib/backend/pgbk/pgbk.go:
```diff
@@ -0,0 +1,580 @@
+
+// setupAndMigrate sets up the database schema, applying the migrations in the
+// [schemas] slice starting from the first non-applied one.
+func (b *Backend) setupAndMigrate(ctx context.Context) error {
```
I see this code is very similar to https://github.com/gravitational/teleport/pull/29057/files#diff-4a9bdc66c1efc75d6078bc6616215a6b6784458b655f395bebcac293e69c852fR267

1. Do we want to keep versioning per table? I assume we need that as backend and events don't need to use Postgres?
2. Shouldn't we make that logic more generic/common?

espadolini replied:
Made the logic generic in 04043cc1cffa246aa6e4404488bfe02d92030a65.

We should use different table names for pgbk and pgevents because if they're pointed to the same database one of the two will assume that the schema is already correct even though tables are missing, so we'll get very strange errors after a supposedly successful initialization.

jakule commented on lib/backend/pgbk/pgbk.go:
```diff
@@ -0,0 +1,580 @@
+	}
+
+	if migrateErr != nil {
+		return migrateErr
```
```suggestion
		return trace.Wrap(migrateErr)
```

espadolini replied:
but it was already a trace error 😭

fixed in 04043cc1cffa246aa6e4404488bfe02d92030a65

jakule commented on lib/backend/pgbk/pgbk.go:
```diff
@@ -0,0 +1,580 @@
+func (b *Backend) CloseWatchers() { b.buf.Clear() }
+
+// Clock implements implements [backend.Backend].
+func (b *Backend) Clock() clockwork.Clock { return clockwork.NewRealClock() }
```
Can you comment on why this is always the real clock? I assume that it's because the backend is also using the DB time, so mocking would not work.

espadolini replied:
641c39cfca

espadolini replied:
in a previous version we shipped the local fake clock time to the database on every transaction, but using the server time has the significant advantage of getting all auth servers to agree about which items exist and which are expired, in spite of any possible clock drift

jakule commented on lib/backend/pgbk/pgbk_test.go:
```diff
@@ -0,0 +1,69 @@
+
+func TestPostgresBackend(t *testing.T) {
+	// recommended: "expiry_interval":"500ms","change_feed_poll_interval":"500ms"
+	paramString := os.Getenv("TELEPORT_PGBK_TEST_PARAMS_JSON")
```
Where is the connection string? In other words, how will this test know where the instance is?

espadolini replied:
Updated comment in e364d78fa710cc9fc82e2755d5bf52689d3dbdc1: those are options that should be added, you are meant to pass a json blob containing all the params including `"conn_string":"..."`.

jakule commented on lib/backend/pgbk/background.go:
```diff
@@ -0,0 +1,261 @@
+				// TODO(espadolini): try other alternatives:
+				// - `DELETE FROM kv WHERE key IN (SELECT ...)`, might be faster
+				//   or slower depending on the batch size
+				// - `WITH to_delete AS (SELECT ...) DELETE FROM ... USING to_delete WHERE key = to_delete.key`
```
This sounds the most "natural" to me. Have you also though about creating a temp table?

espadolini replied:
The performance is certainly good enough with all these options; I've tweaked it slightly in 8fd585fa67b4b610d342bc202e0f40a0285cde5a to select keys in expiration order, which seems to have slightly better performance when combined with the LIMIT (it uses the index on expires rather than doing a seq scan which might take a while before it gets enough rows - the planner can't know how many will be selected). We can change it later if the behavior under heavy real load shows that the expiry deletion ends up forcing too many serialization retries; for instance, I suspect that grabbing a connection, doing a `CREATE TEMP TABLE to_delete AS SELECT key FROM kv WHERE (item is expired)` in read-only, deferred mode and then later doing a `DELETE FROM kv WHERE (item is expired and key is in to_delete)` might be better with regards to messing the least with the rest of the backend operations.

jakule reviewed (commented):
First pass

espadolini committed:
clarify error messages

espadolini committed:
clarify TELEPORT_PGBK_TEST_PARAMS_JSON

espadolini committed:
Describe why we always use the real clock

espadolini committed:
move SetupAndMigrate to common

espadolini committed:
update revision comment

espadolini committed:
clarify batches

espadolini committed:
actually use the context in conn.Close

espadolini committed:
give a reason why we silence logs in the change feed

espadolini commented on lib/backend/pgbk/background.go:
```diff
@@ -0,0 +1,261 @@
+	defer func() {
+		ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
+		defer cancel()
+		if err := conn.Close(context.Background()); err != nil && ctx.Err() != nil {
```
It is! 😳 fixed in 32dc569333

espadolini commented on lib/backend/pgbk/pgbk.go:
```diff
@@ -0,0 +1,580 @@
+					Key:     key,
+					Value:   value,
+					Expires: time.Time(expires).UTC(),
+					// Revision: crockfordLowercaseBase32.EncodeToString(revision.Bytes[:]),
```
c4a8b9d940

espadolini committed:
order expiration deletes by expiry time

espadolini committed:
Merge branch 'master' into espadolini/pgbk

espadolini committed:
reduce defaultChangeFeedBatchSize down to 1000

we seem to be handling outrageous load fine even with 1000, and 10k (if
actually hit) all but guarantees that we go over the default work_mem,
forcing postgres to spill to disk; it's a tunable, anyway

espadolini committed:
parse expires as timestamptz in pollChangeFeed

espadolini committed:
Merge branch 'master' into espadolini/pgbk

espadolini committed:
implement GetName

espadolini committed:
blindly try to grant REPLICATION

espadolini committed:
default_transaction_isolation in RuntimeParams

espadolini committed:
Always use NewDefaultAzureCredential

espadolini committed:
retry unique_violation and exclusion_violation

espadolini committed:
Merge branch 'master' into espadolini/pgbk

espadolini committed:
fix pgevents.TestConfig

jakule commented on lib/backend/pgbk/background.go:
```diff
@@ -147,6 +146,16 @@ func (b *Backend) runChangeFeed(ctx context.Context) error {
+	// REPLICATION attribute yet
+	// HACK(espadolini): ALTER ROLE CURRENT_USER REPLICATION just crashes postgres on Azure
+	if _, err := conn.Exec(ctx,
+		fmt.Sprintf("ALTER ROLE \"%v\" REPLICATION", poolConfig.ConnConfig.User),
```
Any chances for the `User` to be empty?

espadolini replied:
It's legal to have an empty username (just very ill-advised); here it'll just result in a harmless `ERROR:  zero-length delimited identifier at or near """"`. If someone really wants to use the default `postgres` user by running a blank one then they'll just have to deal with not having Teleport enable replication automatically.

jakule commented on lib/backend/pgbk/common/utils.go:
```diff
@@ -131,7 +131,12 @@ func retry[T any](ctx context.Context, log logrus.FieldLogger, isIdempotent bool
 
-		if pgErr != nil && (pgErr.Code == pgerrcode.SerializationFailure || pgErr.Code == pgerrcode.DeadlockDetected) {
+		if pgErr != nil &&
+			(pgErr.Code == pgerrcode.SerializationFailure ||
+				pgErr.Code == pgerrcode.DeadlockDetected ||
+				pgErr.Code == pgerrcode.UniqueViolation ||
+				pgErr.Code == pgerrcode.ExclusionViolation) {
```
Can you move this condition to a function and give it a name? `IsViolentError()`?

espadolini replied:
Extracted in `isSerializationErrorCode` in 37a971c166

jakule commented on lib/backend/pgbk/pgbk.go:
```diff
@@ -159,16 +154,18 @@ func NewWithConfig(ctx context.Context, cfg Config) (*Backend, error) {
-	poolConfig.AfterConnect = func(ctx context.Context, c *pgx.Conn) error {
-		_, err := c.Exec(ctx, "SET default_transaction_isolation TO serializable", pgx.QueryExecModeExec)
-		return trace.Wrap(err)
+	if poolConfig.ConnConfig.RuntimeParams["default_transaction_isolation"] != "" {
```
nit: `default_transaction_isolation` is repeated here 4x. Can you create a constant or use one if it's available somewhere?

espadolini replied:
polished in 238e29e8f8

jakule commented on lib/backend/pgbk/pgbk.go:
```diff
@@ -159,16 +154,18 @@ func NewWithConfig(ctx context.Context, cfg Config) (*Backend, error) {
+		log.WithField("default_transaction_isolation", poolConfig.ConnConfig.RuntimeParams["default_transaction_isolation"]).
+			Warn("The default_transaction_isolation parameter was overridden in the connection string; proceeding with an unsupported configuration.")
+	} else {
+		poolConfig.ConnConfig.RuntimeParams["default_transaction_isolation"] = "serializable"
```
nit: 
```suggestion
		poolConfig.ConnConfig.RuntimeParams["default_transaction_isolation"] = pgx.Serializable
```

espadolini replied:
I disagree with that, `pgx.Serializable` only happens to be the `"serializable"` string (behind a newtype, so it would have to be `string(pgx.Serializable)` in our case) for internal `pgx` convenience - it's used for string formatting/concatenation in places, internally - but it wouldn't be a compatibility break for `pgx` to change it, as it's only intended to be passed in `pgx.TxOptions.IsoLevel`.

espadolini committed:
extract isSerializationErrorCode

espadolini committed:
polish the default_transaction_isolation check

espadolini commented on lib/backend/pgbk/background.go:
```diff
@@ -147,6 +146,16 @@ func (b *Backend) runChangeFeed(ctx context.Context) error {
 
+	// this can be useful if we're some sort of admin but we haven't gotten the
+	// REPLICATION attribute yet
+	// HACK(espadolini): ALTER ROLE CURRENT_USER REPLICATION just crashes postgres on Azure
```
Will do as soon as I figure out where to file an issue in Azure for reference 😅

espadolini committed:
Fix typo in Clock doc comment

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

espadolini committed:
expires is a Timestamptz

espadolini committed:
Merge branch 'master' into espadolini/pgbk
